mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/TaskExecutor.java

```diff
@@ -0,0 +1,650 @@
+
+        // if somehow we are not able to register with the resource manager or something else errors out
+        // while trying to connect to the resource manager, then we expect the startup to fail.
+//    return serviceStartedFuture.thenApply(dontCare -> cxn);
+    }
+
+    private ExecutorService getIOExecutor() {
+        return this.ioExecutor;
+    }
+
+    private CompletableFuture<TaskExecutorReport> getCurrentReport(Time timeout) {
+        return callAsync(() -> {
+            if (this.currentTask == null) {
+                return TaskExecutorReport.available();
+            } else {
+                return TaskExecutorReport.occupied(currentTask.getWorkerId());
+            }
+        }, timeout);
+    }
+
+    @Slf4j
+    @ToString(of = "gateway")
+    @RequiredArgsConstructor
+    static class ResourceManagerGatewayCxn extends AbstractScheduledService {
+
+        private final int idx;
+        private final TaskExecutorRegistration taskExecutorRegistration;
+        private final ResourceClusterGateway gateway;
+        private final Time heartBeatInterval;
+        private final Time heartBeatTimeout;
+        private final Time timeout = Time.of(1000, TimeUnit.MILLISECONDS);
+        private final Function<Time, CompletableFuture<TaskExecutorReport>> currentReportSupplier;
+        private final int tolerableConsecutiveHeartbeatFailures;
+
+        private int numFailedHeartbeats = 0;
+
+        @Override
+        protected String serviceName() {
+            return "ResourceManagerGatewayCxn-" + String.valueOf(idx);
+        }
+
+        @Override
+        protected Scheduler scheduler() {
+            return Scheduler.newFixedDelaySchedule(
+                0,
+                heartBeatInterval.getSize(),
+                heartBeatInterval.getUnit());
+        }
+
+        @Override
+        public void startUp() throws Exception {
+            log.info("Trying to register with resource manager {}", gateway);
+            try {
+                gateway
+                    .registerTaskExecutor(taskExecutorRegistration)
+                    .get(heartBeatTimeout.getSize(), heartBeatTimeout.getUnit());
+            } catch (Exception e) {
+                // the registration may or may not have succeeded. Since we don't know let's just
+                // do the disconnection just to be safe.
+                log.error("Registration to gateway {} has failed; Disconnecting now to be safe", gateway,
+                    e);
+                try {
+                    gateway.disconnectTaskExecutor(
+                            new TaskExecutorDisconnection(taskExecutorRegistration.getTaskExecutorID(),
+                                taskExecutorRegistration.getClusterID()))
+                        .get(2 * heartBeatTimeout.getSize(), heartBeatTimeout.getUnit());
+                } catch (Exception inner) {
+                    log.error("Disconnection has also failed", inner);
+                }
+                throw e;
```

Comment:
Is the handling policy here to crash and restart the whole process on each request failure? Shall we consider some retries?

Reply:
No, the process does not restart. When the resource manager gateway connection fails to establish, the service goes from the "RUNNING" to the "FAILED" state. The task executor at this point (which is listening to these state transitions) would try to reestablish a new connection with the resource manager. So, the process doesn't restart.